mitmproxy/addons/upload.py

```diff
@@ -0,0 +1,124 @@
+import os.path
+import typing
+import random
+import datetime
+import time
+from urllib.request import Request, urlopen
+from urllib.error import HTTPError, URLError
+import mimetypes
+
+from mitmproxy import command
+from mitmproxy import exceptions
+from mitmproxy import flowfilter
+from mitmproxy import io
+from mitmproxy import ctx
+from mitmproxy import flow
+
+import mitmproxy.types
+
+from mitmproxy.addons import multipart
+
+
+class Upload:
+    def __init__(self):
+        self.stream = None
+        self.filt = None
+        self.active_flows = set()  # type: Set[flow.Flow]
+
+    def open_file(self, path):
+        mode = "wb"
+        path = os.path.expanduser(path)
+        return open(path, mode)
+
+    def read_file(self, path):
+        path = os.path.expanduser(path)
+        with open(path, "rb") as f:
+            content = f.read()
+        return content
+
+    def start_stream_to_path(self, path, flt):
+        try:
+            f = self.open_file(path)
+        except IOError as v:
+            raise exceptions.OptionsError(str(v))
+        self.stream = io.FilteredFlowWriter(f, flt)
+        self.active_flows = set()
+
+    def configure(self, updated):
+        # We're already streaming - stop the previous stream and restart
+        if "save_stream_filter" in updated:
+            if ctx.options.save_stream_filter:
+                self.filt = flowfilter.parse(ctx.options.save_stream_filter)
+                if not self.filt:
+                    raise exceptions.OptionsError(
+                        "Invalid filter specification: %s" % ctx.options.save_stream_filter
+                    )
+            else:
+                self.filt = None
+        if "save_stream_file" in updated or "save_stream_filter" in updated:
+            if self.stream:
+                self.done()
+            if ctx.options.save_stream_file:
+                self.start_stream_to_path(ctx.options.save_stream_file, self.filt)
+
+    def base36encode(self, integer):
+        chars, encoded = "0123456789abcdefghijklmnopqrstuvwxyz", ""
+
+        while integer > 0:
+            integer, remainder = divmod(integer, 36)
+            encoded = chars[remainder] + encoded
+
+        return encoded
+
+    @command.command("upload.file")
+    def upload(self, flows: typing.Sequence[flow.Flow]) -> None:
+
+        d = datetime.datetime.utcnow()
+        id = self.base36encode(int(time.mktime(d.timetuple()) * 1000 * random.random()))[0:7]
+        try:
+            f = self.open_file(id)  # Making temporary file to store the flows
```

Comment:
As mentioned in https://github.com/mitmproxy/mitmproxy/issues/2779#issuecomment-358385276, you can pass a BytesIO object to FlowWriter. No need to write anything to disk.

Reply:
Okay thanks